lib/apiservers/engine/backends/container.go

```diff
@@ -359,123 +451,167 @@ func (c *ContainerBackend) ContainerExecStart(ctx context.Context, eid string, s
-		}
+		var err error
+		ec, err = c.taskStartHelper(op, id, eid, name)
+		return err
+	}
+
+	if err := retry.DoWithConfig(operation, engerr.IsConflictError, backoffConf); err != nil {
+		op.Errorf("Failed to start Exec task for container(%s) due to error (%s)", id, err)
+		return err
+	}
+
+	attach := ec.OpenStdin || ec.OpenStdout || ec.OpenStderr
+
+	handle, err := c.Handle(id, name)
+	if err != nil {
+		op.Errorf("Failed to obtain handle during exec start for container(%s) due to error: %s", id, err)
+	}
+
+	// we need to be able to cancel it
+	taskOp, cancel := trace.WithCancel(&op, "exec task wait on %s", eid)
+	defer cancel()
+
+	attachResult := make(chan error, 1)
+	startResult := make(chan error, 1)
 
-		ec, err := c.containerProxy.InspectTask(op, handle, eid, id)
+	go func() {
+		// wait for the task to start
+		err := c.containerProxy.WaitTask(op, handle, name, eid)
 		if err != nil {
-			return err
+			startResult <- err
 		}
 
-		handle, err = c.containerProxy.BindTask(op, handle, eid)
+		// TODO: if wait task can return an up-to-date handle we will not need this additional call
+		handle, err := c.Handle(id, name)
 		if err != nil {
-			return err
+			startResult <- err
 		}
 
-		// exec doesn't have separate attach path so we will decide whether we need interaction/runblocking or not
-		attach := ec.OpenStdin || ec.OpenStdout || ec.OpenStderr
-		if attach {
-			handle, err = c.containerProxy.BindInteraction(ctx, handle, name, eid)
-			if err != nil {
-				op.Errorf("Failed to initiate interactivity during exec start for container(%s) due to error: %s", id, err)
-				return err
-			}
+		ec, err = c.containerProxy.InspectTask(op, handle, eid, id)
+		if err != nil {
+			startResult <- err
 		}
 
-		if err := c.containerProxy.CommitContainerHandle(ctx, handle, name, 0); err != nil {
-			op.Errorf("Failed to commit handle for container(%s) due to error: %s", id, err)
-			return err
+		log.Debugf("task start: %#v, %#v", ec, ec.ProcessConfig)
+		if ec != nil && ec.ProcessConfig != nil && ec.ProcessConfig.ErrorMsg != "" {
+			startResult <- errors.New(ec.ProcessConfig.ErrorMsg)
 		}
 
-		// we need to be able to cancel it
-		taskCtx, cancel := trace.WithCancel(&op, "exec task wait on %s", eid)
-		defer cancel()
+		close(startResult)
+	}()
 
-		// we do not return an error here if this fails. TODO: Investigate what exactly happens on error here...
-		// FIXME: This being in a retry could lead to multiple writes to stdout(?)
+	if attach {
 		go func() {
-			defer trace.End(trace.Begin(eid))
+			attachResult <- c.attachHelper(taskOp, ec, stdin, stdout, stderr, false)
+			close(attachResult)
+		}()
+	}
 
-			// wait property collector
-			if err := c.containerProxy.WaitTask(taskCtx, id, name, eid); err != nil {
-				op.Errorf("Task wait returned %s, canceling the context", err)
+	// wait for either wait to succeed, fail, or for stream connection to error out
+	select {
+	case err := <-attachResult:
```

Comment:
This is really very minor, but I was expecting to find `startResult` handling before `attachResult` in this select, as that's the order they occur in the preceding goroutine.